### PR TITLE
[FSTSI-229] - Implementation of Characters

### PR DESCRIPTION
**Pull request by neilnahid1**
This is the pull request for the implementation of the characters in my game. 
Some of the notable changes I made is refactored my code.
Added a family of class called Elements which is a property of the Unit class.
The elements have effects on the damage output of a unit based on his element and the enemies' element.
Also added a family of class called Race which is also a property of the Unit class.
The Race class will determine the unit's damage and health.
I've also implemented a simple battle system for this one in order to simulate a battle between two units and to be able to test the functionality of different classes.

### CODE REVIEW

File: classes/Elements/Corruption.php [hunk 1]
```diff
@@ -2,13 +2,34 @@
...
+            0.50, # fire resistance percent
+            0.50, # wind resistance percent
+            1.0   # corruption resistance percent
+        );
     }
-    public function dialog()
+    /**
+     * returns the appropriate damage based on defenders resistance to this element
+     */
+    public function applyElementResistance(\Classes\Units\Unit $attacker, \Classes\Units\Unit $defender) : int
```
Comment (lcabornay) [inline classes/Elements/Corruption.php]: @neilnahid1 you can use 
```
// importing or use
use \Classes\Units\Unit;
...

   public function applyElementResistance(Unit $attacker, Unit $defender) : int

```

File: main.php [hunk 2]
```diff
@@ -13,8 +13,11 @@
 main();
 function main()
 {
-    $hero = UnitBuilder::buildHero("Elf", "Basic", "Neil", 1);
+    //initialise the hero and a sample unit
+    $hero = UnitBuilder::buildHero("Elf", "Physical", "Neil", 1);
```
Comment (lcabornay) [inline main.php:17]: for string constant, you can create variable.
```
const heroType = 'Elf';
const damageType = 'Physical';
const name = 'Neil';
```
retrieve using
```
self::heroType
```
```
$hero = UnitBuilder::buildHero(self::heroType, self::damageType, self::name, 1);
```

**Review by lcabornay — CHANGES_REQUESTED** [review]
@neilnahid1 kindly check my review comments and. thanks.

File: classes/Elements/Corruption.php (hunk 1 above)
Comment (neilnahid1) [reply classes/Elements/Corruption.php]: Noted sir.

File: main.php (hunk 2 above)
Comment (neilnahid1) [reply main.php:17]: Okay sir, will do.

**Comment by neilnahid1** [comment]
Will edit the PR request to be more descriptive.

**Pull request closed without merging**